### DISCUSSION TITLE
Is it possible to add a secret with username and uri metadata at the same time?

### DISCUSSION THREAD

**Discussion by cyberautomate** [marshallwp/SecretManagement.Warden#7] [General]
I ahve tried the code below and get this error: Set-Secret: Unable to add secret test to vault bitwarden
```
$data = @{
	UserName = "myname"
	uris = "https://www.test.com"
}

Set-Secret -Name test -Secret 'adfasdf' -Metadata $data
```

bitwarden is the name of the vault. I can login to the vault no problem and Test-SecretVault comes back as True

The only way I can get Set-Secret to work is with this workflow.
```
set-secret
cmdlet Set-Secret at command pipeline position 1
Supply values for the following parameters:
Name: testing
SecureStringSecret: ********
Is this SecureString a UserName, Password, TOTP, URIs, or SecureNote?: password
```

What am I missing?

**Comment by marshallwp** [comment]
Not sure how I missed this until now 😳.  But better late than never I suppose.  The bottom line is that this module doesn't currently support the `Metadata` parameter.  Note that [support of that is optional](https://github.com/PowerShell/SecretManagement?tab=readme-ov-file#secret-metadata) for extensions, and Bitwarden doesn't really have a concept of such as a thing separate from the secret.

What this module _does_ do is infer secret type based on the datatype of the new secret and in some cases it's contents.
https://github.com/marshallwp/SecretManagement.Warden/blob/bbb0a449c3cf5905d0e30485cfce64a71531a082/SecretManagement.Warden.Extension/public/Set-Secret.ps1#L21-L60

For example:
- `PSCredentials` are treated as a logins.
- `HashTables` are tested for field names and treated as a Login, Card, Identity or SecureNote depending on what is found.
- `Strings` and `SecureStrings` don't include enough information to make the determination and so result in the prompt you see above.

Existing secrets will always use the same datatype, only new ones have it inferred like this.

If you want to create a Secret as a one-liner, you should put all the fields into a `HashTable` and pass that as the secret value.  So given your example above:

```pwsh
# NOTE: The HashTable can contain a Username and Password of the SecureString data type
#       instead of plain text, but the latter is clearer for this demonstrative.
$data = @{
	UserName = "myname"
        Password = "SuperSecure"
	uris = @("https://www.test.com")
}

Set-Secret -Name test -Secret $data
```
Or at least that's how it _should_ work.  The uris bit doesn't seem to be getting stored for some reason, which annoys me as I have code just for that:
https://github.com/marshallwp/SecretManagement.Warden/blob/bbb0a449c3cf5905d0e30485cfce64a71531a082/SecretManagement.Warden.Extension/public/Set-Secret.ps1#L86

Unfortunately it seems I [overlooked that field](https://github.com/marshallwp/SecretManagement.Warden/blob/bbb0a449c3cf5905d0e30485cfce64a71531a082/SecretManagement.Warden.Extension/test/public/Set-Secret.Tests.ps1#L131-L136) when creating tests, so it never got flagged. 😩  Will look into why that's occurring a bit later.

Again, apologies for the long response time, and thank you for bringing stuff like this to my attention!  I don't use the full breadth of the module myself, so it's not always apparent to me when something isn't working as expected or when documentation needs to be clarified.

**Comment by marshallwp** [comment]
This has been fixed.  The latest version published to PSGallery and released here will now accept new/existing URIs.  Provide them as either a string array or an array of hashtables with `uri` and `match` properties.

### Example
```pwsh
$data = @{
	UserName = "myname"
        Password = "SuperSecure"
	uris = @("https://www.test.com")
}
```
OR
```pwsh
$data = @{
	UserName = "myname"
        Password = "SuperSecure"
	uris = @(
            @{
                uri = "https://www.test.com"
                match = "host"
             }
        )
}
```
